src/overmapbuffer.cpp

```diff
@@ -715,6 +715,86 @@ tripoint overmapbuffer::find_closest( const tripoint &origin, const std::string
+{
+    // Check the origin before searching adjacent tiles!
+    if( allow_subtype_matches
+        ? check_ot_subtype( type, origin.x, origin.y, origin.z )
+        : check_ot_type( type, origin.x, origin.y, origin.z ) ) {
+        if( visited_point.find( origin ) == visited_point.end() ) {
+            return origin;
+        }
+    }
+
+    // Function overloading for NPC pathfinding
+    // visited_point is added to check points NPC has been, if any of them matches function ignores the location
+
+    int max = ( radius == 0 ? OMAPX * 5 : radius );
+    const int z = origin.z;
+    // expanding box
+    for( int dist = 0; dist <= max; dist++ ) {
+        // each edge length is 2*dist-2, because corners belong to one edge
+        // south is +y, north is -y
+        for( int i = 0; i < dist * 2; i++ ) {
+            //start at northwest, scan north edge
+            int x = origin.x - dist + i;
+            int y = origin.y - dist;
+            if( allow_subtype_matches
+                ? check_ot_subtype( type, x, y, z )
+                : check_ot_type( type, x, y, z ) ) {
+                if( !must_be_seen || seen( x, y, z ) ) {
+                    if( visited_point.find( tripoint( x, y, z ) ) == visited_point.end() ) {
+                        return tripoint( x, y, z );
+                    }
+                }
+            }
+
+            //start at southeast, scan south
+            x = origin.x + dist - i;
+            y = origin.y + dist;
+            if( allow_subtype_matches
+                ? check_ot_subtype( type, x, y, z )
+                : check_ot_type( type, x, y, z ) ) {
+                if( !must_be_seen || seen( x, y, z ) ) {
+                    if( visited_point.find( tripoint( x, y, z ) ) == visited_point.end() ) {
+                        return tripoint( x, y, z );
+                    }
+                }
+            }
+
+            //start at southwest, scan west
+            x = origin.x - dist;
+            y = origin.y + dist - i;
+            if( allow_subtype_matches
+                ? check_ot_subtype( type, x, y, z )
+                : check_ot_type( type, x, y, z ) ) {
+                if( !must_be_seen || seen( x, y, z ) ) {
+                    if( visited_point.find( tripoint( x, y, z ) ) == visited_point.end() ) {
+                        return tripoint( x, y, z );
+                    }
+                }
+            }
+
+            //start at northeast, scan east
+            x = origin.x + dist;
+            y = origin.y - dist + i;
+            if( allow_subtype_matches
+                ? check_ot_subtype( type, x, y, z )
+                : check_ot_type( type, x, y, z ) ) {
+                if( !must_be_seen || seen( x, y, z ) ) {
+                    if( visited_point.find( tripoint( x, y, z ) ) == visited_point.end() ) {
+                        return tripoint( x, y, z );
+                    }
+                }
+            }
+        }
+    }
+    return overmap::invalid_tripoint;
```

Comment:
And don't return *any* tripoint at all when you don't have one. Return `cata::nullopt` instead. The return type `optional<T>` indicates to the caller that the function may not return any value at all. The caller must therefor check the return value.

Returning just a `tripoint` does not tell the caller that the function can return an "invalid value". One has to add documentation (which is never done, and not read anyway).

Reply:
As I said above, the function was copied for function overloading. I guess developer who made it did it.

Reply:
I hesitate returning other than invaild_tripoint (which is tripoint( INT_MIN, INT_MIN, INT_MIN )) because there are no_goal_point in npc class which is just a tripoint( INT_MIN, INT_MIN, INT_MIN), and it is widely used in npcmove.cpp and probably more.

Reply:
But it's you changing now tripoint to cata:optional<tripoint>.

Reply:
I know, just thinking how to deal with without having any troubles.

Reply:
I should revert it to tripoint, and we should have another branch to fix them all.

Reply:
Wait, you *copied* the function? Why? Just to add the `visited_point` set? So now we have two *nearly* identical functions. Please don't do that. Algorithms should be implemented in one place only.

Suggestion: add the `visited_point` parameter to the original function (and add the checks using it there). For existing calls to it that don't supply the set, add an overload that *forward* to the new functions like this:
```
tripoint find_closese( ..., const set &visited_point )
{
    ... // code using visited_point
}

tripoint find_closese( ... )
{
    static const set &empty_visited_point;
    return find_closese( ..., empty_visited_point ); // forward all parameters
}
```

Reply:
@BevapDin That's what I didn't think. really appreciated!